service/interpreter/workflowImpl.go

```diff
@@ -782,10 +794,74 @@ func processStateExecution(
 		stateExeId, state, stateExecutionLocal, commandReq,
 		completedTimerCmds, completedSignalCmds, completedInterStateChannelCmds,
 	)
+
+	// Wait for decider trigger (ANY/ALL command completed) OR continue-as-new threshold
 	_ = provider.Await(ctx, func() bool {
 		return IsDeciderTriggerConditionMet(commandReq, completedTimerCmds, completedSignalCmds, completedInterStateChannelCmds) || continueAsNewCounter.IsThresholdMet()
 	})
+
 	commandReqDoneOrCanceled = true
+
+	// Wait for command threads to drain. After the command request await completes, command threads
+	// may still be in the process of storing retrieved data into completedXXXCmds maps.
+	// We must wait for these threads to finish before assembling command results, otherwise
+	// retrieved data will be lost (the thread retrieved it but never stored it before we read the maps).
+	// We only wait for threads of completed commands (those with entries in completedXXXCmds),
+	// not all threads, which preserves ANY_COMMAND_COMPLETED semantics.
+	if globalVersioner.IsAfterVersionOfWaitingCommandThreads() {
+		if err := provider.Await(ctx, func() bool {
```

Comment:
Why not just :
```
	if err := provider.Await(ctx, func() bool {
		for _, completed := range waitForThreadNames {
			if !completed {
				return false
			}
		}
		return true
	}); err != nil {
		return nil, service.WaitingCommandsStateExecutionStatus, err
	}
```
I don't fully follow the logic here.

Reply:
I got too in the weeds of thinking about what needed to happen. Your logic is much simpler. commandReqDoneOrCanceled = true does what I was trying to accomplish.